File: backend/08-review/src/main/kotlin/com/dokandar/review/observability/Metrics.kt
package com.dokandar.review.observability

import io.prometheus.client.CollectorRegistry
import io.prometheus.client.Counter
import io.prometheus.client.Gauge
import io.prometheus.client.Histogram
import io.prometheus.client.exporter.common.TextFormat
import java.io.StringWriter

object Metrics {
    const val SVC = "08-review"
    val registry: CollectorRegistry = CollectorRegistry.defaultRegistry
    val httpRequests: Counter = Counter.build("http_requests_total", "HTTP requests").labelNames("method", "route", "status").register()
    val httpDuration: Histogram = Histogram.build("http_request_duration_seconds", "HTTP latency").labelNames("method", "route").register()
    val reviewPosts: Counter = Counter.build("review_posts_total", "review posts").labelNames("service", "result").register()
    val reviewVotes: Counter = Counter.build("review_votes_total", "review votes").labelNames("service").register()
    val reviewReports: Counter = Counter.build("review_reports_total", "review reports").labelNames("service", "reason").register()
    val reviewHidden: Counter = Counter.build("review_hidden_total", "auto-hides").labelNames("service").register()
    val reviewAggUpdates: Counter = Counter.build("review_aggregate_updates_total", "rating_aggregates upserts").labelNames("service").register()
    val reviewKafka: Counter = Counter.build("review_kafka_messages_total", "projector messages").labelNames("service", "topic", "result").register()
    val outboxPublished: Counter = Counter.build("review_outbox_published_total", "outbox rows published").labelNames("service").register()
    val outboxPending: Gauge = Gauge.build("review_outbox_pending", "unsent outbox rows").labelNames("service").register()

    fun render(): String { val w = StringWriter(); TextFormat.write004(w, registry.metricFamilySamples()); return w.toString() }
}

File: backend/07-coupon/src/Observability/Metrics.cs
using Prometheus;

namespace Coupon.Observability;

public static class Metrics
{
    public const string Svc = "07-coupon";
    public static readonly Counter HttpRequests = Prometheus.Metrics.CreateCounter(
        "http_requests_total", "HTTP requests", new CounterConfiguration { LabelNames = new[] { "method", "route", "status" } });
    public static readonly Histogram HttpDuration = Prometheus.Metrics.CreateHistogram(
        "http_request_duration_seconds", "HTTP latency", new HistogramConfiguration { LabelNames = new[] { "method", "route" } });
    public static readonly Counter CouponValidations = Prometheus.Metrics.CreateCounter(
        "coupon_validations_total", "ValidateCoupon outcomes", new CounterConfiguration { LabelNames = new[] { "service", "result" } });
    public static readonly Histogram CouponValidationMs = Prometheus.Metrics.CreateHistogram(
        "coupon_validation_ms", "ValidateCoupon latency (ms)", new HistogramConfiguration { LabelNames = new[] { "service" }, Buckets = new[] { 1, 5, 10, 25, 50, 100, 250, 500.0 } });
    public static readonly Counter CouponDrafts = Prometheus.Metrics.CreateCounter(
        "coupon_drafts_total", "coupons drafted", new CounterConfiguration { LabelNames = new[] { "service", "scope" } });
    public static readonly Counter CouponApprovals = Prometheus.Metrics.CreateCounter(
        "coupon_approvals_total", "four-eyes approvals", new CounterConfiguration { LabelNames = new[] { "service" } });
    public static readonly Counter CouponRevocations = Prometheus.Metrics.CreateCounter(
        "coupon_revocations_total", "coupon revocations", new CounterConfiguration { LabelNames = new[] { "service" } });
    public static readonly Counter CouponRedemptions = Prometheus.Metrics.CreateCounter(
        "coupon_redemptions_total", "redemptions recorded", new CounterConfiguration { LabelNames = new[] { "service" } });
    public static readonly Counter OutboxPublished = Prometheus.Metrics.CreateCounter(
        "coupon_outbox_published_total", "outbox rows published to Kafka", new CounterConfiguration { LabelNames = new[] { "service" } });
    public static readonly Gauge OutboxPending = Prometheus.Metrics.CreateGauge(
        "coupon_outbox_pending", "unsent outbox rows", new GaugeConfiguration { LabelNames = new[] { "service" } });
}
